### ISSUE TITLE
AccessControl related methods from IO

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#9872]
```
M:System.IO.Directory.SetAccessControl(System.String,System.Security.AccessControl.DirectorySecurity)
M:System.IO.DirectoryInfo.Create(System.Security.AccessControl.DirectorySecurity)
M:System.IO.DirectoryInfo.CreateSubdirectory(System.String,System.Security.AccessControl.DirectorySecurity)
M:System.IO.DirectoryInfo.GetAccessControl
M:System.IO.DirectoryInfo.GetAccessControl(System.Security.AccessControl.AccessControlSections)
M:System.IO.DirectoryInfo.SetAccessControl(System.Security.AccessControl.DirectorySecurity)
M:System.IO.Directory.CreateDirectory(System.String,System.Security.AccessControl.DirectorySecurity)
M:System.IO.Directory.GetAccessControl(System.String)
M:System.IO.Directory.GetAccessControl(System.String,System.Security.AccessControl.AccessControlSections)
M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess)
M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean)
M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean,System.Int32)
M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean,System.Int32,System.Boolean)
```

from #9465

**Comment by ianhays** [comment]
We've got at least a few of these already in [System.IO.FileSystem.AccessControl](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs). The issue with adding others (e.g. the FIleStream constructors) is that they're implemented as extensions to FileSystem rather than being built in to Filesystem. 

We may need to consider merging the FileSystem.AccessControl assembly into FileSystem and just throw PlatformNotSupportedException on Unix.

**Comment by danmosemsft** [comment]
@ianhays Talk to Wes about whether/how much the access control members will be in .NET Standard 2.0.

**Comment by ianhays** [comment]
These were all excluded from netstandard in https://github.com/dotnet/standard/pull/21.

Discussion in https://github.com/dotnet/standard/issues/15

Some of these APIs are still available in the windows specific [System.IO.FileSystem.AccessControl assembly](https://github.com/dotnet/corefx/tree/master/src/System.IO.FileSystem.AccessControl), but they are not a part of netstandard.

**Comment by ianhays** [comment]
cc: @weshaggard

**Comment by bbarry** [comment]
will these apis be available as one of the “Windows Desktop Packs” mentioned at BUILD for Core 3?

**Comment by JeremyKuhne** [comment]
> will these apis be available as one of the “Windows Desktop Packs” mentioned at BUILD for Core 3?

Don't know yet. Do you have a scenario where moving to the methods in System.IO.FileSystem.AccessControl is difficult or impossible?

**Comment by bbarry** [comment]
We have a large application base that utilizes [SAP Adaptive Server Enterprise](https://www.sap.com/products/sybase-ase.html) (formerly Sybase) via their provided ADO.NET driver and we could potentially switch to Core for several libraries as well as future development (already have an AspNetCore MVC application targeting desktop framework). 3rd party drivers don't work properly due to certain datatypes widely used in our schema outside of the defined TDS protocol for parameter sizes and the ODBC driver historically fails to work every time I try to convert to it. I've been unable to get any vendor support for the driver for several years now (for example in getting a correctly working async implementation or fixes for other issues we have had to work around) and have little hope for a native .NET Core driver implementation. 

Anyway the driver immediately fails to open a connection with an exception:

```
Unhandled Exception: System.TypeInitializationException: The type initializer for 'Sybase.Data.AseClient1.AseConnection' threw an exception. ---> System.MissingMethodException: Method not found: 'System.Security.AccessControl.DirectorySecurity System.IO.Directory.GetAccessControl(System.String)'.
   at Sybase.Data.AseClient1.AseConnection.LoadLibraries()
   at Sybase.Data.AseClient1.AseConnection..cctor()
   --- End of inner exception stack trace ---
   at Sybase.Data.AseClient1.AseConnection..ctor(AseConnection realConnection, String connectionString)
   at Sybase.Data.AseClient.AseConnection..ctor(String connectionString)
```

when built for Core framework for obvious reasons.

**Comment by JeremyKuhne** [comment]
@bbarry Thanks for the details! It really helps to get input like this when we're planning. Reopening for consideration- we don't have the right milestones yet, putting in Future.

Doing this would require moving the implementation into System.IO.FileSystem as explained above.

**Comment by ocalvo** [comment]
BTW: This breaks Enter-VsShell module in Visual Studio

**Comment by JeremyKuhne** [comment]
@carlossanlop This should be addressed by your security extensions, can you validate?

**Comment by carlossanlop** [comment]
Here are the APIs we have already merged:

Shipped in 3.1:

#### FileInfo that creates a FileStream
- API Proposal: https://github.com/dotnet/corefx/issues/41614
- PR: https://github.com/dotnet/corefx/pull/42099
- Available in 3.1: https://github.com/dotnet/corefx/pull/42228
```cs
namespace System.IO
{
    public static class FileSystemAclExtensions
    {
        public static System.IO.FileStream Create(this System.IO.FileInfo fileInfo, System.IO.FileMode mode, System.Security.AccessControl.FileSystemRights rights, System.IO.FileShare share, int bufferSize, System.IO.FileOptions options, System.Security.AccessControl.FileSecurity fileSecurity);
    }
 }
```

### DirectoryInfo that creates a directory
- API Proposal: https://github.com/dotnet/corefx/issues/41614
- PR: https://github.com/dotnet/corefx/pull/41834
- Available in 3.1: https://github.com/dotnet/corefx/pull/42041/files
```cs
namespace System.IO
{
    public static class FileSystemAclExtensions
    {
        public static void Create(this System.IO.DirectoryInfo directoryInfo, System.Security.AccessControl.DirectorySecurity directorySecurity);
    }
 }      
```

Will be shipped in 5.0:

#### Mutex
- API Proposal: https://github.com/dotnet/corefx/issues/41662
- PR: https://github.com/dotnet/corefx/pull/42281
```cs
namespace System.Threading
{
    public static class SemaphoreAcl
    {
        public static System.Threading.Mutex Create(bool initiallyOwned, string name, out bool createdNew, System.Security.AccessControl.MutexSecurity mutexSecurity);
    }
}
```

#### Semaphore
- API Proposal: https://github.com/dotnet/corefx/issues/41662
- PR: https://github.com/dotnet/corefx/pull/42377
```cs
namespace System.Threading
{
    public static class SemaphoreAcl
    {
        public static System.Threading.Semaphore Create(int initialCount, int maximumCount, string name, out bool createdNew, System.Security.AccessControl.SemaphoreSecurity semaphoreSecurity);
    }
}
```

#### EventWaitHandle
- API Proposal: https://github.com/dotnet/corefx/issues/41662
- PR: https://github.com/dotnet/corefx/pull/42213
```cs
namespace System.Threading
{
    public static class EventWaitHandleAcl
    {
        public static System.Threading.EventWaitHandle Create(bool initialState, System.Threading.EventResetMode mode, string name, out bool createdNew, System.Security.AccessControl.EventWaitHandleSecurity eventSecurity);
    }
}
```

#### AnonymousPipeServerStream
- API Proposal: https://github.com/dotnet/corefx/issues/41657
- PR: https://github.com/dotnet/runtime/pull/111

```cs
namespace System.IO.Pipes
{
    public static class AnonymousPipeServerStreamAcl
    {
        public static System.IO.Pipes.AnonymousPipeServerStream Create(System.IO.Pipes.PipeDirection direction, System.IO.HandleInheritability inheritability, int bufferSize, System.IO.Pipes.PipeSecurity pipeSecurity);
    }
}
```

#### NamedPipeServerStream
- API Proposal: https://github.com/dotnet/corefx/issues/41657
- PR: https://github.com/dotnet/runtime/pull/317
```cs
namespace System.IO.Pipes
{
    public static class NamedPipeServerStreamAcl
    {
        public static System.IO.Pipes.NamedPipeServerStream Create(string pipeName, System.IO.Pipes.PipeDirection direction, int maxNumberOfServerInstances, System.IO.Pipes.PipeTransmissionMode transmissionMode, System.IO.Pipes.PipeOptions options, int inBufferSize, int outBufferSize, System.IO.Pipes.PipeSecurity pipeSecurity, System.IO.HandleInheritability inheritability, System.IO.Pipes.PipeAccessRights additionalAccessRights);
    }
}
```

#### MemoryMappedFile (Work in progress)
- API Proposal: https://github.com/dotnet/runtime/issues/941
- PR: https://github.com/dotnet/runtime/pull/465
```cs
namespace System.IO.MemoryMappedFiles
{
    public class MemoryMappedFileSecurity : System.Security.AccessControl.ObjectSecurity<System.IO.MemoryMappedFiles.MemoryMappedFileRights>
    {
        public MemoryMappedFileSecurity();
    }
    public static class MemoryMappedFileAcl
    {
        public static System.IO.MemoryMappedFiles.MemoryMappedFile CreateFromFile(System.IO.FileStream fileStream, string mapName, long capacity, System.IO.MemoryMappedFiles.MemoryMappedFileAccess access, System.IO.MemoryMappedFiles.MemoryMappedFileSecurity memoryMappedFileSecurity, System.IO.HandleInheritability inheritability, bool leaveOpen);
        public static System.IO.MemoryMappedFiles.MemoryMappedFile CreateNew(string mapName, long capacity, System.IO.MemoryMappedFiles.MemoryMappedFileAccess access, System.IO.MemoryMappedFiles.MemoryMappedFileOptions options, System.IO.MemoryMappedFiles.MemoryMappedFileSecurity memoryMappedFileSecurity, System.IO.HandleInheritability inheritability);
        public static System.IO.MemoryMappedFiles.MemoryMappedFile CreateOrOpen(string mapName, long capacity, System.IO.MemoryMappedFiles.MemoryMappedFileAccess access, System.IO.MemoryMappedFiles.MemoryMappedFileOptions options, System.IO.MemoryMappedFiles.MemoryMappedFileSecurity memoryMappedFileSecurity, System.IO.HandleInheritability inheritability);
        public static System.IO.MemoryMappedFiles.MemoryMappedFileSecurity GetAccessControl(this System.IO.MemoryMappedFiles.MemoryMappedFile memoryMappedFile);
        public static void SetAccessControl(this System.IO.MemoryMappedFiles.MemoryMappedFile memoryMappedFile, System.IO.MemoryMappedFiles.MemoryMappedFileSecurity memoryMappedFileSecurity);
    }
}
```

#### DirectorySecurity that creates a DirectoryInfo (Pending)
- API Proposal: https://github.com/dotnet/corefx/issues/8895
```cs
namespace System.IO
{
    public static class FileSystemAclExtensions
    {
        public static DirectoryInfo CreateDirectory(this System.Security.AccessControl.DirectorySecurity directorySecurity, string path);
    }
}
```

#### Notes
From the list shared by Dan at the beginning of this issue, these 4 APIs will not be ported:

>_M:System.IO.DirectoryInfo.CreateSubdirectory(System.String,System.Security.AccessControl.DirectorySecurity)_

This can be addressed using the `CreateDirectory` extension method for `DirectorySecurity`.

>_M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess)_
>_M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean)_
>_M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean,System.Int32)_
>_M:System.IO.FileStream.#ctor(System.IntPtr,System.IO.FileAccess,System.Boolean,System.Int32,System.Boolean)_

All these can be addressed with the `Create` extension method for `FileInfo` that returns a `FileStream`.
